Feature: Add External Source (and External Event) Attribute Validation

pranav-super opened the pull request:
This pull request adds external source/event attribute validation to Aerie.

This means that when an external source is uploaded to AERIE, via endpoints like the CLI or the UI, it is forwarded here, has its formatting checked against a schema, and additionally has the attributes of the source and each of its events checked for validity, before submitting it to the database.

Submission of event/source type schemas for the purpose of later validation is also handled here - the gateway simply checks that they are valid JSON Schema. 

A more thorough explanation of the feature can be found in the AERIE PR, [here](https://github.com/NASA-AMMOS/aerie/pull/1606).

JosephVolosin committed:
Add routes for CreateExternalSource, CreateExternalSourceType, and CreateExternalEventType

Commit:
require source type to specify associated event types

Commit:
check event types in external source

Commit:
ensure external event attributes actually checked

Commit:
add external source validation to gateway

Commit:
add validation tests

Commit:
improve return

Commit:
additional error handling

Commit:
remove allowed event type functionality from source type creation and validation

Commit:
final edits for error handling

Commit:
remove console.logs

JosephVolosin committed:
Update error responses from external source/event end-points

Mythicaeda commented on src/packages/external-event/external-event.ts:
```diff
@@ -0,0 +1,136 @@
+  } catch (error) {
+    logger.error((error as Error).message);
+    res.status(500);
+    res.send((error as Error).message);
```
This is **not** Aerie Gateway standard. Do **not** just send a text message back as an error message, you need to send a JSON. See L43 of `middleware.ts` for an example of the correct way to return an error status. This applies to all places where you're returning error statuses.

Mythicaeda commented on src/packages/external-event/external-event.ts:
```diff
@@ -0,0 +1,136 @@
+async function uploadExternalEventType(req: Request, res: Response) {
+  logger.info(`POST /uploadExternalEventType: Entering function...`);
+  const authorizationHeader = req.get('authorization');
+  logger.info(`POST /uploadExternalEventType: ${authorizationHeader}`);
```
Remove this, you should never be logging user credentials like this.

JosephVolosin replied:
Sorry - missed this from cleaning up some testing debugs, removed now!

Mythicaeda commented on src/packages/external-event/external-event.ts:
```diff
@@ -0,0 +1,136 @@
+const ajv = new Ajv();
+
+async function uploadExternalEventType(req: Request, res: Response) {
+  logger.info(`POST /uploadExternalEventType: Entering function...`);
```
This log is unnecessary, please remove it.

Mythicaeda commented on src/packages/external-event/external-event.ts:
```diff
@@ -0,0 +1,136 @@
+  // Validate schema is valid JSON Schema
+  try {
+    const schemaIsValid: boolean = ajv.validateSchema(attribute_schema);
+    if (!schemaIsValid) {
+      throw new Error("Schema was not a valid JSON Schema.");
+    }
+  } catch (error) {
+    logger.error((error as Error).message);
+    res.status(500);
+    res.send((error as Error).message);
```
Can `ajv.validateSchema` throw an Error? If not, why is this a try-catch? If it can, then this comment can be resolved.

JosephVolosin replied:
Good catch - no it looks like all errors are just directly accessible through `ajv.errors`. I updated the `logger` and `res.send` statements to use `ajv.errors` instead of relying on a `try / catch`.

JosephVolosin committed:
Update logging messages & use JSON for all responses

JosephVolosin committed:
Improve error handling w. AJV.errors

JosephVolosin committed:
Formatting fix

Mythicaeda commented on src/packages/external-event/external-event.ts:
```diff
@@ -0,0 +1,136 @@
+  }
+
+  // Make sure name in schema (title) and provided name match
+  try {
+    if (attribute_schema['title'] === undefined || attribute_schema.title !== external_event_type_name) {
+      throw new Error('Schema title does not match provided external event type name.');
+    }
+  } catch (error) {
+    logger.error(
+      `POST /uploadExternalEventType: Error occurred during External Event Type ${external_event_type_name} upload`,
+    );
+    logger.error((error as Error).message);
+    res.status(500).send({ message: (error as Error).message });
+    return;
+  }
```
Why is this a try-catch instead of an if-return?

JosephVolosin replied:
Cleaned up, no longer using `try/catch`

Mythicaeda commented on src/packages/external-event/external-event.ts:
```diff
@@ -0,0 +1,136 @@
+    return;
+  }
+
+  logger.info(`POST /uploadExternalEventType: Attribute schema was VALID`);
```
Please use present tense in logs, unless the log is explicitly referring to a "prior state":
```suggestion
  logger.info(`POST /uploadExternalEventType: Attribute schema is VALID`);
```

JosephVolosin replied:
All `logger` and `response` messages now use present tense!

Mythicaeda commented on src/packages/external-event/external-event.ts:
```diff
@@ -0,0 +1,136 @@
+    method: 'POST',
+  });
+
+  type CreateExternalEventTypeResponse = {
```
Types should be declared in a file in the `types` directory, not inline. Additionally, this type should not start at the `data` wrapper object but just be the middle `{ attribute_schema: object; name: string }`

Please use `types/plans.ts` for a reference of what I mean.

JosephVolosin replied:
Moved all type definitions out to `types/external-source.ts`

Mythicaeda commented on src/packages/external-event/external-event.ts:
```diff
@@ -0,0 +1,136 @@
+   *     tags:
+   *       - Hasura
+   */
+  app.post('/uploadExternalEventType', uploadExternalEventType);
```
This endpoint is currently unauthenticated. Example of an authenticated endpoint from L716 of `packages/plan/plan.ts`. Please also speak with @duranb regarding the `upload.single` part.

```ts
app.post('/uploadDataset', upload.single('external_dataset'), refreshLimiter, auth, uploadDataset);
```

JosephVolosin replied:
Re-worked the routes so they properly use `auth` with `plan.ts` as a reference

Mythicaeda commented on src/packages/external-event/external-event.ts:
Organization question: why are these separate `external-event` and `external-source` packages instead of one united package?

JosephVolosin replied:
I started with them separate since they'd been separated at times when doing the UI work in the original pull-request, but I do think it makes sense to condense to one package - I can't think of a name though, maybe just naming it `external-source` or `external-source-events`?

Mythicaeda replied:
Isn't the feature bundled up under the name `external-events` in the UI? If so, that works. Otherwise `external-source` works.

JosephVolosin replied:
Ended up moving all new routes to a singular package as `external-source`!

JosephVolosin committed:
Rework external source & event end-points to use multipart/form-data

JosephVolosin committed:
Rework uploadExternalSource to use form data rather than file

Parsing is done on the UI side, and parsed data is sent as a multipart/form-data request

JosephVolosin committed:
Rework Hasura responses

JosephVolosin committed:
Formatting fixes

Commit:
attempt merge i

Commit:
attempt merge ii

Commit:
update gateway tests

JosephVolosin committed:
Partial refactor of new schema compilation code

Still a few areas to refactor/cleanup, WIP

JosephVolosin committed:
Add description for request objects

Commit:
update for e2e tests

Commit:
minor fix post rebase

JosephVolosin committed:
external_events => events

Commit:
formatting

JosephVolosin committed:
formatting

JosephVolosin committed:
Formatting

JosephVolosin committed:
Formatting, cleanup

duranb reviewed (dismissed):
Can we keep a consistent camel cased variable names where applicable?

duranb commented on test/external_source.validation.test.ts:
```diff
@@ -0,0 +1,406 @@
+import Ajv from 'ajv';
```
can this file be renamed to be kebab cased, unless it needs to be snake?

JosephVolosin replied:
Yep! Renamed

JosephVolosin committed:
Fix name casing

JosephVolosin committed:
e as Error => (e as Error).message

JosephVolosin committed:
Fix improper time regex for schema

JosephVolosin committed:
Move schemas => src/

JosephVolosin committed:
Add ajv to package.json

Mythicaeda commented on src/packages/external-source/external-source.ts:
```diff
@@ -0,0 +1,497 @@
+async function uploadExternalSourceEventTypes(req: Request, res: Response) {
+  const authorizationHeader = req.get('authorization');
+
+  const {
+    headers: { 'x-hasura-role': roleHeader, 'x-hasura-user-id': userHeader },
+  } = req;
+
+  const { body } = req;
+  const { event_types, source_types } = body;
```
This can be done at once:
```suggestion
  const {
    headers: { 'x-hasura-role': roleHeader, 'x-hasura-user-id': userHeader },
    body: {event_types, source_types}
  } = req;
```

This comment applies to multiple functions in this file

JosephVolosin replied:
Implemented!

Mythicaeda commented on src/packages/external-source/external-source.ts:
```diff
@@ -0,0 +1,497 @@
+async function uploadExternalSource(req: Request, res: Response) {
+  const authorizationHeader = req.get('authorization');
+
+  const {
+    headers: { 'x-hasura-role': roleHeader, 'x-hasura-user-id': userHeader },
+  } = req;
+
+  const { body } = req;
```
This can be simplified:
```suggestion
  const {
    headers: { 'x-hasura-role': roleHeader, 'x-hasura-user-id': userHeader },
    body: { source, events }
  } = req;
```

JosephVolosin replied:
Implemented!

Mythicaeda commented on src/packages/external-source/external-source.ts:
```diff
@@ -0,0 +1,497 @@
+
+  const { body } = req;
+
+  if (typeof body !== 'object') {
+    logger.error(
+      `POST /uploadExternalSourceEventTypes: Body of request must be a JSON, with two stringified properties: "source" and "events".`,
+    );
+    res
+      .status(500)
+      .send({ message: `Body of request must be a JSON, with two stringified properties: "source" and "events".` });
+    return;
+  }
```
This `if` statement is unnecessary. Additionally, this error message describes a **different body**  than the swagger docs of the endpoint. Please align the two. (Additionally please move the error string into a const if you're going to keep repeating it)

JosephVolosin replied:
Removed the `if` and fixed the swagger docs for both end-points. Also went through and all `logger` + `res.status(500).send({...})` combos are now using a consistent `errorMsg` for both!

JosephVolosin committed:
Cleanup from comments

JosephVolosin committed:
Lint fix

JosephVolosin committed:
Make GQL_API_URL manually

Commit:
update to allow automatic creation of
types and allow either/or event/source types in schema

Commit:
minor comment

Commit:
update package?

JosephVolosin committed:
Fix linting

Commit:
minor edit to refresh SonarQube

Commit:
reduce code repetition in tests

Commit:
lint tests

JosephVolosin committed:
Move external source parsing to Gateway

JosephVolosin committed:
Add postgres-interval to prod

Required for time parsing in external source validation

JosephVolosin committed:
Receive type schemas as JSON

JosephVolosin committed:
Remove 'Z' before parsing times

JosephVolosin committed:
Formatting fix

JosephVolosin committed:
Remove limit from multer in external-source.ts

JosephVolosin committed:
Use source derivation group field unless request has override

Previously was ignoring the field completely

AaronPlave committed:
Typo

AaronPlave committed:
Comments and variable name tweaks

AaronPlave committed:
Refactor process for handling event types without schema

dandelany commented:
A few updates...
* Reviewed code with @AaronPlave yesterday and found two improvements to make, which @AaronPlave and @JosephVolosin are working on committing to this branch:
  - We improved the logic for checking for events with nonexistent schemas to not need a nested loop
  - There was some confusion over whether or not we needed the "empty schemas" that are saved for events/sources without attributes. @JosephVolosin convinced us they're good to have in the database (for later filtering work), but it seems they are not being used in the "megaschema" in this gateway code and it's not clear if they needed to be. Joe will write a e2e test for this case(s) to check, and will add them to the megaschema if so.
* once these changes are made @AaronPlave and I will do a more exhaustive user test, on track to approve.

dandelany committed:
small external event fixes: allow missing attributes key, fix check for schemas for empty attributes

dandelany commented:
Update - @JosephVolosin and I fixed remaining known issues on this branch related to uploading event sources without attributes or schemas - ready for a final user test with @AaronPlave

dandelany committed:
update parseJSONFile to reject if json is invalid

dandelany committed:
code formatting

dandelany reviewed (approved):
Tested with @AaronPlave and all looks good from Gateway's perspective. We pushed one more additional fix in [e3b0e23](https://github.com/NASA-AMMOS/aerie-gateway/pull/116/commits/e3b0e232213abf767547da0bc9f8e7640de1afcb) because the (shared) json parsing code wasn't correctly throwing an error in the case of invalid JSON - looks good now!

mattdailis commented:
The UI tells me I can leave the derivation group blank, but when I try to upload a source without a derivation group I get an error. Am I doing something wrong?

<img width="448" height="198" alt="Screenshot 2025-07-17 at 2 20 29 PM" src="https://github.com/user-attachments/assets/d2163e7f-495b-4cd9-be0d-96d99db90f17" />


<img width="754" height="212" alt="Screenshot 2025-07-17 at 2 21 07 PM" src="https://github.com/user-attachments/assets/c49336ef-8cd8-4078-81ed-3fb05fa69594" />


At first glance, this GraphQL mutation seems to assume that a derivation group will be provided:
https://github.com/NASA-AMMOS/aerie-gateway/blob/41b7e7f873409d499a59bc21c309d3588415c095/src/packages/external-source/gql.ts#L7-L14

dandelany commented:
For posterity we addressed @mattdailis 's last comment with a [UI message change here](https://github.com/NASA-AMMOS/aerie-ui/pull/1560/commits/df2abb152bb7ea21f6c905b6bfbf65434bdf4f6a).

This is ready to go and doesn't have any breaking changes, merging this ahead of the expected [backend](https://github.com/NASA-AMMOS/aerie/pull/1606) and [UI](https://github.com/NASA-AMMOS/aerie-ui/pull/1560) PRs merging soon.
